feat(cli,toolkit-lib): drift detection via `cdk drift` and `toolkit.drift()`

Leo10Gama opened the pull request:
Added an additional action and command `toolkit.drift()` and `cdk drift`, which invokes drift detection. Prints an output of the drift results listing out any resources in the stack that may have drifted.

Tested via `yarn build && yarn test` and adding in a few integ tests.

Running the command locally shows the following. With one resource drifted:
![image](https://github.com/user-attachments/assets/7db0f6fa-72ec-44b5-91ae-952c6a40e791)

With no resources drifted: (image outdated)
![image](https://github.com/user-attachments/assets/b38cc96a-c404-4339-bc1b-d09311d505f1)

---
By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license

Commit:
Add drift detection to cdk diff

Commit:
Add drift detection to cdk diff

Commit:
Fix output for no drifts

Commit:
Fix output for no drifts

Commit:
Merge

Commit:
Merge branch 'drift' of https://github.com/Leo10Gama/aws-cdk-cli into drift

Commit:
Not sure how all that got there lmao oops

Commit:
Add test for multiple resources

rix0rrr commented:
Thanks! This is a great feature to have!

I understand that we're detecting some kind of "diff" from a user point of view, but I'm very put-off by drift detection being part of `cdk diff`. That's just not what CDK diff historically did, and it's confusing to me. I wonder if it will be confusing to more people?

```
┌────────────┐            ╔═════════════╗            ┌────────────┐
│            │    diff    ║             ║    drift   │            │
│  Template  │◀──────────▶║    Stack    ║◀──────────▶│  Reality   │
│            │            ║             ║            │            │
└────────────┘            ╚═════════════╝            └────────────┘
                                                                   
                   ^                                               
               cdk diff                                            
            does this today                                        
```

Can we make this its own subcommand to begin with? `cdk drift` ?

> Tested via yarn build && yarn test, which both compile.

Surely there's more? 🥹 An integ test would be nice.

rix0rrr commented on packages/@aws-cdk/toolkit-lib/lib/api/deployments/cfn-api.ts:
```diff
@@ -466,6 +468,86 @@ export async function stabilizeStack(
+      throw new ToolkitError(`Drift detection failed: ${response.DetectionStatusReason}`);
+    }
+
+    if (attempts++ > 30) {
+      throw new ToolkitError('Drift detection timed out after 30 attempts');
+    }
```
This is a bit hard to reason about, what with the backoff and all. What I'd recommend is doing a deadline-driven loop.

Something like this:

```ts
const deadline = Date.now() + SOME_TIME_THAT_SEEMS_REASONABLE;
while (true) {
  if (Date.now() > deadline) {
    throw new Error(`Took more than ${SOME_TIME_THAT_SEEMS_REASONABLE}!`);
  }

  // Normal work...
}
```

Advantages:

* Need to keep track of `attempts`
* Very clear what the maximum runtime is
* Error message is more useful to users

rix0rrr commented on packages/@aws-cdk/toolkit-lib/lib/api/deployments/cfn-api.ts:
```diff
@@ -466,6 +468,86 @@ export async function stabilizeStack(
+    }
+
+    // Calculate backoff with jitter
+    const jitter = Math.random() * 1000;
```
Jitter is usually calculated over the entire potentially delay period.

Here are the two different methods of calculating sleep (yours and the standard one):

![image](https://github.com/user-attachments/assets/dd8e5cdd-fe98-4292-ba8d-d3f97467377e)

For the first one (`baseDelay + jitter(1s)`), you will see that as `baseDelay` grows the jitter will become a negligible part of the total sum, so you might as well not have it.

(Since you have a `maxDelay`, also make sure to jittering over that as well)

-----

There is a standard way of calculating exponential backoff. Don't try to outsmart that standard way. 

https://aws.amazon.com/blogs/architecture/exponential-backoff-and-jitter/

Leo10Gama replied:
Good point, and thanks for the quick crash course! I think in conjunction with your above comment (re: keeping track of attempts) that simpler deadline-driven loop approach makes a lot more sense. I can revamp this chunk of the code

rix0rrr commented on packages/@aws-cdk/toolkit-lib/lib/api/diff/diff-formatter.ts:
```diff
@@ -329,6 +363,36 @@ export class DiffFormatter {
     return { formattedDiff, permissionChangeType: this.permissionType() };
   }
+
+  public formatStackDrift(options: FormatStackDriftOptions): FormatStackDriftOutput {
```
Do we get information about resources that don't support drift detection anywhere? 

I wonder if this gives a false sense of confidence over resources that don't support the operation at all, and whether we should tell users that we didn't check 'em.

mrgrain replied:
It's quite limited these days what is not supported, but yes.

Leo10Gama replied:
Closest thing I'm able to find about resources without drift detection support is [this doc](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/resource-import-supported-resources.html). I tried testing locally with one of these values (specifically `AWS::EC2::SubnetRouteTableAssociation` since it's only listed to not support drift detection) and in both the console and the command return, the resource itself just doesn't come up:
![image](https://github.com/user-attachments/assets/84592ed7-7cce-4ded-af36-640327fee66b)
![image](https://github.com/user-attachments/assets/877d10d9-18b0-45b9-bb41-b28b32469874)


If we want to show users which resources we haven't checked, we can probably cross-reference the output list with the list of resources we can see in the stack, but since drift detection already doesn't check mention them, it might make things more confusing for users. I'd opt to leave the messages be without noting those resources

rix0rrr replied:
It's not that hard, right?

If we get the list of all resources, and **remove** those that are `IN_SYNC` (or drifted), then we are left with the list of resources that we don't know? And we can print that list? 

Also, I wouldn't show the `IN_SYNC` resources at all. If the question is `cdk diff`, then the answer "here are 99 resources that are a-ok" isn't that interesting.

Leo10Gama replied:
Agreed, I'm not arguing that it's difficult to implement, I just don't see the reasoning why we'd want to mention those resources. If drift detection already ignores these resources, it'd be weird for us to highlight them, since it could lead to the confusion of "why does the console say I have no drift but CDK says some resources weren't checked for drift?" 

Happy to add that implementation, I just want to understand better why this is something customers will want to see, as opposed to something that may clutter the output and make customers think there's something they need to fix, when in some cases (like the `AWS::EC2::SubnetRouteTableAssociation`) they can't even touch those resources.

Re: the `IN_SYNC` resources, the reason I had the output show all resources in the above image was for debugging purposes to show that the SDK call doesn't return the `AWS::EC2::SubnetRouteTableAssociation`. Final implementation will of course not contain these, I should've mentioned that with the image 😅

Leo10Gama replied:
I see your point, yup. I'll whip up some changes to include that logic

mrgrain commented on packages/@aws-cdk/cloudformation-diff/lib/format.ts:
```diff
@@ -19,6 +20,16 @@ export interface FormatStream extends NodeJS.WritableStream {
+/**
+ * Resource drift status types from CloudFormation
+ */
+export enum ResourceDriftStatus {
+  MODIFIED = 'MODIFIED',
+  DELETED = 'DELETED',
+  NOT_CHECKED = 'NOT_CHECKED',
+  IN_SYNC = 'IN_SYNC',
+}
```
We prefer discriminated string unions for internal types these days. Also the enum seems unused outside this file, so no need to export

```suggestion
type ResourceDriftStatus = 'MODIFIED' | 'DELETED' | 'NOT_CHECKED' | 'IN_SYNC';
```

Leo10Gama replied:
Actually, looking further into it, we can probably just import the `StackResourceDriftStatus` directly from the SDK instead of redefining it here, which might be cleaner still

mrgrain commented on packages/@aws-cdk/toolkit-lib/lib/api/aws-auth/sdk.ts:
```diff
@@ -88,6 +88,14 @@ import type {
   UpdateTerminationProtectionCommandInput,
   UpdateTerminationProtectionCommandOutput,
   StackSummary,
+  DescribeStackDriftDetectionStatusCommandInput,
+  DescribeStackDriftDetectionStatusCommandOutput,
+  DescribeStackResourceDriftsCommandOutput,
+  DetectStackDriftCommandInput,
+  DetectStackDriftCommandOutput,
+  DetectStackResourceDriftCommandInput,
+  DetectStackResourceDriftCommandOutput,
+  DescribeStackResourceDriftsCommandInput,
```
These surely will require new permissions? If not can you make a note why? Or are you preparing a corresponding change to the bootstrap stack?

Leo10Gama replied:
They will need new permissions, yes. Specifically, after looking into it, it seems I'll need to add these:
```
cloudformation:DescribeStackDriftDetectionStatus
cloudformation:DescribeStackResourceDrifts
cloudformation:DetectStackDrift
cloudformation:DetectStackResourceDrift
```
I am preparing to make those changes to the PR as well to update the bootstrap stack.

Leo10Gama replied:
Testing this locally, it looks like I don't need to add these permissions to give access for drift detection. Do you know what role `cdk diff` is using to execute these changes? It may be that it either has these permissions already or is using the admin role.

rix0rrr replied:
I think `cdk diff` might be using the CLI's credentials, or perhaps the lookup role.

Leo10Gama replied:
Confirmed with AppSec, this uses the lookup role, which has the following permissions:
```yml
...
"cloudformation:Describe*",
"cloudformation:Detect*",
...
```
Hence, no bootstrap changes are necessary.

mrgrain commented on packages/@aws-cdk/toolkit-lib/lib/api/deployments/cfn-api.ts:
```diff
@@ -466,6 +468,86 @@ export async function stabilizeStack(
+): Promise<DescribeStackDriftDetectionStatusCommandOutput | undefined> {
+  await ioHelper.notify(IO.DEFAULT_TOOLKIT_DEBUG.msg(format('Waiting for drift detection %s to complete...', driftDetectionId)));
+
+  const maxDelay = 30_000; // 30 seconds max delay
```
This is an awwful long time without any visual status update.

mrgrain commented on packages/@aws-cdk/toolkit-lib/lib/api/diff/diff-formatter.ts:
```diff
@@ -329,6 +363,36 @@ export class DiffFormatter {
+    const stream = new StringWriteStream();
+    let driftCount = 0;
+
+    if (!this.driftResults?.StackResourceDrifts) {
+      return { formattedDrift: '', numResourcesWithDrift: 0 };
+    }
```
These seems incorrect. It's not `0` resources with drift - it's unkown.

Leo10Gama replied:
Good point. The chunk is supposed to catch if the drift results are empty when the method is called, which right now doesn't seem like it will happen with my implementation, but I agree it would return 0 when really drift detection hasn't run. Would it be better to return a more descriptive message, maybe with `numResourcesWithDrift: -1`?

mrgrain replied:
Yeah, that kind of highlights a weakness in the approach of introducing an optional prop into the formatter class constructor, to later maybe call the format function.

As far as I can tell, there is no reason this needs to be part of `DiffFormatter`. So let's just make it it's own file and api: `drift/formatter.ts` or whatever. You can probably also simplify how the class handles the formatting since the detour via `stream` is just an artifact of legacy code. Instead pushing onto an array and joining with newlines will do the same.

mrgrain commented on packages/@aws-cdk/toolkit-lib/lib/api/diff/diff-formatter.ts:
```diff
@@ -60,10 +77,15 @@ interface DiffFormatterProps {
+  /**
+   * The results of stack drift
+   */
+  readonly driftResults?: DescribeStackResourceDriftsCommandOutput;
```
why not in template info?

Leo10Gama replied:
I had debated on including it in there vs outside, and my reasoning was that it wasn't as closely related to the template diff things being computed to warrant grouping them. Though going through and refactoring it now, I can see how combining them might make more sense in the grand scheme of things, and might make some of the logic in the toolkit-lib simpler as well, since it seems like that's meant to return the template info only

mrgrain commented on packages/aws-cdk/lib/cli/cdk-toolkit.ts:
```diff
@@ -264,6 +258,19 @@ export class CdkToolkit {
+        let driftResults = undefined;
+
+        if (options.detectDrift) {
+          const env = await this.props.deployments.resolveEnvironment(stack);
+          const cfn = (await this.props.sdkProvider.forEnvironment(env, Mode.ForReading)).sdk.cloudFormation();
+
+          driftResults = await detectStackDrift(
+            cfn,
+            asIoHelper(this.ioHost, 'diff'),
+            stack.stackName,
+          );
```
This typically goes into a helper that receives `this.props.deployments` and `this.props.sdkProvider`

mrgrain reviewed (changes requested):
Also needs integ tests and added to `toolkit-lib` (see Rico's comments).

If we are adding this as a new action/command, the action code should live in toolkit-lib and CLI just be calling it.

GavinZZ commented:
> Thanks! This is a great feature to have!
> 
> I understand that we're detecting some kind of "diff" from a user point of view, but I'm very put-off by drift detection being part of `cdk diff`. That's just not what CDK diff historically did, and it's confusing to me. I wonder if it will be confusing to more people?
> 
> ```
> ┌────────────┐            ╔═════════════╗            ┌────────────┐
> │            │    diff    ║             ║    drift   │            │
> │  Template  │◀──────────▶║    Stack    ║◀──────────▶│  Reality   │
> │            │            ║             ║            │            │
> └────────────┘            ╚═════════════╝            └────────────┘
>                                                                    
>                    ^                                               
>                cdk diff                                            
>             does this today                                        
> ```
> 
> Can we make this its own subcommand to begin with? `cdk drift` ?

Let's discuss this further during Tuesday's meeting. While I'm open to making this a separate command, I'd like to explain my original reasoning for including it in `cdk diff`:

I envisioned `cdk diff` as a broader capability for detecting any differences affecting a deployed stack, whether those differences:
- Come from changes in local CDK code
- Result from manual modifications to deployed resources through console or CLI (Drifts)

However, I recognize your point about the historical context of `cdk diff` and the potential for user confusion. This would be a good topic for us to explore during the meeting, weighing the benefits of command consolidation against clarity and user expectations.

Commit:
Move drift logic to helper function

Commit:
Duplicate into toolkit-lib

Commit:
Remove ResourceDriftStatus enum
In favour of using the SDK's enum value, which CFN uses anyways

Commit:
Minor tweaks

Commit:
Update timeout mechanism

Commit:
Move driftResults to be within TemplateInfo

Commit:
Change message when driftResults is undefined

Commit:
Merge branch 'main' into drift

Commit:
Merge hell has been traversed

GavinZZ commented:
I think having a separate command like `cdk drift` and support `cdk diff --drift` meanwhile make sense to me. @Leo10Gama Just a note that if we're going to make drift a standalone CLI command, I think we also need to have a separate AppSec review and Talos engagement for the new CLI command.

mrgrain commented:
@Leo10Gama FYI new actions should primarily be implemented in `toolkit-lib` and than called from the CLI. See the refactor command for an example.

Commit:
Move drift to its own command

Commit:
Merge branch 'main' into drift

Leo10Gama commented:
Still implementing an integ test but pushing changes thus far

Commit:
Move cfn-api methods to api/drift

Commit:
Make numResourcesDrifted optional to simplify output

Commit:
Among other things, added integration tests

Leo10Gama committed:
Merge branch 'main' into drift

Commit:
Include verbose optioning

Commit:
Intermediate merge commit

Commit:
I
ntermediate merge commit

Commit:
Merged from main

Commit:
Merge and add test coverage

Leo10Gama committed:
Merge branch 'main' into drift

Leo10Gama committed:
Merge branch 'main' into drift

Commit:
Remove stack name from error message
Serves to protect potentially sensitive information in the output
from being intercepted.

Commit:
Merge branch 'drift' of https://github.com/Leo10Gama/aws-cdk-cli into drift

Commit:
Fix bug where verboe message appeared in output always

Commit:
Update README.md

rix0rrr commented on packages/@aws-cdk/toolkit-lib/lib/api/drift/drift.ts:
```diff
@@ -0,0 +1,76 @@
+): Promise<DescribeStackDriftDetectionStatusCommandOutput | undefined> {
+  await ioHelper.notify(IO.DEFAULT_TOOLKIT_DEBUG.msg(format('Waiting for drift detection %s to complete...', driftDetectionId)));
+
+  const timeout = 10_000; // if takes longer than 10s, fail
```
Is it realistic for a drift detection operation to finish in 10s? This seems short to me.

I think we probably want a max timeout of a minute or more, and periodically print something like `Still checking...`. We could even print

```
Checking resource drift (5/20)
```

Maybe the timeout even needs to be configurable; it's going to be the first change request coming in on this I'd imagine.

Leo10Gama replied:
Good point, I'll see if I can include some way to have it track resource count as they're being checked, or just something to let the user know the command is still churning away

rix0rrr commented on packages/aws-cdk/lib/cli/cdk-toolkit.ts:
```diff
@@ -675,6 +676,48 @@ export class CdkToolkit {
+  /**
+   * Detect infrastructure drift for the given stack(s)
+   */
+  public async drift(options: DriftCommandOptions): Promise<number> {
```
I think we'd want this code to live in `@aws-cdk/toolkit-lib`, and then call it from the CLI.

Right, @mrgrain ?

mrgrain replied:
Correct. See `refactor()` for how to do this correctly.

Commit:
Add intermediary message while drift detection running

mrgrain commented on packages/@aws-cdk/cloudformation-diff/lib/format.ts:
```diff
@@ -1,9 +1,10 @@
 import * as chalk from 'chalk';
 import type { DifferenceCollection, TemplateDiff } from './diff/types';
 import { deepEqual } from './diff/util';
-import type { Difference, ResourceDifference } from './diff-template';
```
This seems unintentional.

mrgrain commented on .projenrc.ts:
```diff
@@ -431,6 +431,7 @@ const cloudFormationDiff = configureProject(
     deps: [
       '@aws-cdk/aws-service-spec',
       '@aws-cdk/service-spec-types',
+      '@aws-sdk/client-cloudformation',
```
If this is now a runtime dep, than it cannot be a peer dep anymore.

mrgrain commented on packages/@aws-cdk/toolkit-lib/lib/index.ts:
```diff
@@ -14,6 +14,7 @@ export * from './payloads';
 // Supporting acts
 export * from './api/aws-auth';
 export * from './api/cloud-assembly';
+export * from './api/drift';
```
These are not public APIs, let's not export them

```suggestion
```

mrgrain commented on packages/@aws-cdk/toolkit-lib/lib/api/drift/drift-formatter.ts:
```diff
@@ -0,0 +1,146 @@
+/**
+ * Properties specific to formatting the stack drift
+ */
+export interface FormatStackDriftOptions {
+  /**
+   * Silences 'No drift detected' messages
+   *
+   * @default false
+   */
+  readonly quiet?: boolean;
+
+  /**
+   * Whether to be verbose
+   *
+   * @default false
+   */
+  readonly verbose?: boolean;
+}
```
I'm aware this is just copying what `diff` currently does for legacy reasons. But I don't want to carry this forward to a brand new API. If you are really keen on having this, let's get on a call together to figure out a solution. Otherwise just drop this please.

Leo10Gama replied:
Yup, was mainly just copying over the flags and functionality from`diff`. I'm not particularly attached to them if that's not the direction we want for the new API, so I'm happy to drop them. Though in the case of `verbose`, I do want to keep that functionality of showing all resources, including those unchecked (addressing [this concern](https://github.com/aws/aws-cdk-cli/pull/442#discussion_r2081366824)) but maybe to prevent overloading the flag, we can rename `--verbose` to `--show-unchecked` in this command, or something along those lines?

mrgrain commented on packages/@aws-cdk/toolkit-lib/lib/api/drift/interfaces.ts:
```diff
@@ -0,0 +1,34 @@
+/**
+ * Interface for IO operations used in drift detection
+ */
+export interface IDriftIoHelper {
+  /**
+   * Send a notification
+   */
+  notify(msg: any): Promise<void>;
+}
```
I don't see the need for this being different than the regular IoHelper. AI gone rouge?

mrgrain replied:
This change will become clearer once the function is moved to the `toolkit-lib` module. We no longer conditionally emit output. Instead, all output is emitted as messages with an appropriate log level. If you need to support additional output when running `cdk drift -v`, you can assign those messages a `debug` log level.

Your code appears to have 4 sections, which I would classify like this:
- unchanged -> debug
- unchecked -> debug
- modified -> result
- deleted -> result

mrgrain commented on packages/@aws-cdk/cloudformation-diff/lib/format.ts:
```diff
@@ -67,6 +68,89 @@ export function formatSecurityChanges(
+ * @param verbose Whether to output more verbose text (include undrifted resources)
+ * @param logicalToPathMap A map from logical ID to construct path
+ */
+export function formatStackDriftChanges(
```
I don't think this function should live in this package since it's not used in `@aws-cdk/cloudformation-diff` it at all. We can move this to `toolkit-lib`.  WDYT @rix0rrr ?

rix0rrr replied:
Fair point. I guess this now lives with the other formatting code, but I do in general like scoping code to the smallest scope possible. Which means a private function in `toolkit-lib` is better than a shared function in `cloudformation-diff` (that's only intended to be used from `toolkit-lib`).

Commit:
Move things around and into toolkit lib

Commit:
CloudFormation client no longer peer dep

Commit:
De-flag-ify the command

Commit:
Fix formatting

Commit:
Merge with main

Commit:
FIXED

Commit:
Update README.md

Leo10Gama committed:
Merge branch 'main' into drift

Commit:
Update docstring
I realize that this phrasing might be confusing

Commit:
Cleanup commented code

Leo10Gama commented:
AppSec review signed off

mrgrain commented on .projenrc.ts:
```diff
@@ -810,7 +808,7 @@ const toolkitLib = configureProject(
     tsJestOptions: {
       transformOptions: {
-        isolatedModules: false, // we use the respective tsc setting
+        isolatedModules: true,
```
pls revert

mrgrain commented on packages/@aws-cdk/cloudformation-diff/lib/format.ts:
```diff
@@ -2,7 +2,7 @@ import { format } from 'util';
 import type { DifferenceCollection, TemplateDiff } from './diff/types';
 import { deepEqual } from './diff/util';
-import type { Difference, ResourceDifference } from './diff-template';
+import type { ResourceDifference, Difference } from './diff-template';
```
pls revert

mrgrain commented on .projenrc.ts:
```diff
@@ -426,12 +426,10 @@ const cloudFormationDiff = configureProject(
-    peerDeps: [
-      '@aws-sdk/client-cloudformation@^3',
-    ],
     deps: [
       '@aws-cdk/aws-service-spec',
       '@aws-cdk/service-spec-types',
+      '@aws-sdk/client-cloudformation',
```
this change and removal from peerDep is not needed anymore. pls revert

mrgrain commented on packages/@aws-cdk-testing/cli-integ/tests/cli-integ-tests/cdk-cdk-drift---fail-throws-when-drift-is-detected.integtest.ts:
please move these tests to a directory called `drift` (see other tests)

mrgrain commented on packages/@aws-cdk/toolkit-lib/lib/actions/drift/index.ts:
```diff
@@ -0,0 +1,81 @@
+   */
+  readonly stacks: StackSelector;
+
+  /**
+   * Whether to fail with exit code 1 if drift is detected
+   *
+   * @default false
+   */
+  readonly fail?: boolean;
```
This is not a toolkit-lib option, but a CLI feature. please remove here.

mrgrain commented on packages/@aws-cdk/toolkit-lib/lib/actions/drift/index.ts:
```diff
@@ -0,0 +1,81 @@
+  readonly finalResult?: string;
+}
+
+export interface DriftResult {
+  /**
+   * Number of resources with drift
+   */
+  readonly numResourcesWithDrift?: number;
+
+  /**
+   * How many resources were not checked for drift
+   */
+  readonly numResourcesUnchecked?: number;
+
+  /**
+   * Complete formatted drift
+   */
+  readonly formattedDrift: FormattedDrift;
+}
+
+export interface DriftCommandResult {
```
Looks like these are the same, so let's just use `DriftResult`

mrgrain replied:
Okay I see now `DriftResult` is an internal type and `DriftCommandResult` is public. 

Make sure `DriftResult` lives with the formatter (and probably rename it slightly). 

For the public type `DriftResult` is the better name.

mrgrain commented on packages/@aws-cdk/toolkit-lib/lib/api/drift/drift.ts:
```diff
@@ -0,0 +1,87 @@
+  ioHelper: IoHelper,
+  stackName: string,
+): Promise<DescribeStackResourceDriftsCommandOutput> {
+  await ioHelper.notify(IO.CDK_TOOLKIT_I4506.msg(format('Starting drift detection for stack %s...', stackName)));
+
+  // Start drift detection
+  const driftDetection = await cfn.detectStackDrift({
+    StackName: stackName,
+  });
+
+  await ioHelper.notify(IO.CDK_TOOLKIT_I4506.msg(format('Detecting drift for stack %s...', stackName)));
```
Is this good UX? Seems like we are just printing a non actionable message twice. And as soon as we are hitting `waitForDriftDetection` we are sending a third message that's also just "Waiting for"

I'd suggest to keep the first message but use the wording if the second.

mrgrain replied:
Actually the second message be turned into a debug containing the StackDriftDetectionId

mrgrain commented on packages/@aws-cdk/toolkit-lib/lib/api/drift/drift.ts:
```diff
@@ -0,0 +1,87 @@
+  ioHelper: IoHelper,
+  driftDetectionId: string,
+): Promise<DescribeStackDriftDetectionStatusCommandOutput | undefined> {
+  await ioHelper.notify(IO.CDK_TOOLKIT_I4506.msg(format('Waiting for drift detection %s to complete...', driftDetectionId)));
```
Can we skip the initial waiting message? We just printed a message that we are starting the drift.

mrgrain replied:
Is the driftDetectionId useful here? I feel like this would be better as a single debug or trace message right after we get this.

mrgrain commented on packages/@aws-cdk/toolkit-lib/lib/api/drift/drift.ts:
```diff
@@ -0,0 +1,87 @@
+  ioHelper: IoHelper,
+  stackName: string,
+): Promise<DescribeStackResourceDriftsCommandOutput> {
+  await ioHelper.notify(IO.CDK_TOOLKIT_I4506.msg(format('Starting drift detection for stack %s...', stackName)));
```
Please either include a useful payload here (maybe `StackSelectionDetails`?) or make it a default message.

mrgrain commented on packages/@aws-cdk/toolkit-lib/lib/api/drift/drift.ts:
```diff
@@ -0,0 +1,87 @@
+  ioHelper: IoHelper,
+  driftDetectionId: string,
+): Promise<DescribeStackDriftDetectionStatusCommandOutput | undefined> {
+  await ioHelper.notify(IO.CDK_TOOLKIT_I4506.msg(format('Waiting for drift detection %s to complete...', driftDetectionId)));
```
Make this a default message pls

mrgrain commented on packages/@aws-cdk/toolkit-lib/lib/api/drift/drift.ts:
```diff
@@ -0,0 +1,87 @@
+    }
+
+    if (Date.now() > checkIn) {
+      await ioHelper.notify(IO.CDK_TOOLKIT_I4506.msg('Waiting for drift detection to complete...'));
```
make this a default message

mrgrain commented on packages/@aws-cdk/toolkit-lib/lib/api/drift/drift.ts:
```diff
@@ -0,0 +1,87 @@
+      await ioHelper.notify(IO.CDK_TOOLKIT_I4506.msg('Waiting for drift detection to complete...'));
+      checkIn = Date.now() + timeBetweenOutputs;
+    }
+    await new Promise(resolve => setTimeout(resolve, 500));
```
this seems like it's doing nothing?

mrgrain commented on packages/@aws-cdk/toolkit-lib/test/api/drift/drift.test.ts:
```diff
@@ -0,0 +1,640 @@
+import { ToolkitError } from '../../../lib/toolkit/toolkit-error';
+import { mockCloudFormationClient, MockSdk } from '../../_helpers/mock-sdk';
+
+jest.mock('../../../lib/api/io/private', () => {
+  const originalModule = jest.requireActual('../../../lib/api/io/private');
+  return {
+    ...originalModule,
+    IO: {
+      ...originalModule.IO,
+      DEFAULT_TOOLKIT_DEBUG: {
+        msg: jest.fn().mockReturnValue('mocked-message'),
+      },
+    },
+  };
+});
```
There is absolutely no need to mock this module. Look for `TestIoHost` if you need to assert messages.

mrgrain commented on packages/@aws-cdk/toolkit-lib/test/api/drift/drift.test.ts:
```diff
@@ -0,0 +1,640 @@
+});
+
+describe('formatStackDrift', () => {
+  let mockIoHelper: IoHelper;
+  let mockNewTemplate: cxapi.CloudFormationStackArtifact;
```
Checkout `TestIoHost` for a much easier way of mocking and testing.

mrgrain commented on packages/aws-cdk/lib/cli/cli-config.ts:
```diff
@@ -339,6 +339,16 @@ export async function makeConfig(): Promise<CliConfig> {
+          variadic: true,
+        },
+        options: {
+          fail: { type: 'boolean', desc: 'Fail with exit code 1 if drift is detected' },
```
please add a default

mrgrain commented on packages/@aws-cdk/toolkit-lib/lib/toolkit/toolkit.ts:
```diff
@@ -375,6 +378,85 @@ export class Toolkit extends CloudAssemblySourceBuilder {
+    const output = [];
+
+    for (const stack of stacks.stackArtifacts) {
+      const cfn = (await (await this.sdkProvider('drift')).forEnvironment(stack.environment, Mode.ForReading)).sdk.cloudFormation();
```
Can this be outside the loop?

mrgrain replied:
maybe just add a comment pointing out that `stack.environment` might be different

mrgrain commented on packages/@aws-cdk/toolkit-lib/lib/toolkit/toolkit.ts:
```diff
@@ -375,6 +378,85 @@ export class Toolkit extends CloudAssemblySourceBuilder {
+      });
+
+      if (!driftResults) {
+        await ioHelper.notify(IO.CDK_TOOLKIT_I4505.msg(`${chalk.bold(stack.displayName)}: unable to detect drift`));
```
make this a default message if we really need it

mrgrain commented on packages/@aws-cdk/toolkit-lib/lib/toolkit/toolkit.ts:
```diff
@@ -375,6 +378,85 @@ export class Toolkit extends CloudAssemblySourceBuilder {
+      }
+
+      const formatter = new DriftFormatter({
+        ioHelper: asIoHelper(this.ioHost, 'drift'),
```
You already have one
```suggestion
        ioHelper,
```

mrgrain commented on packages/@aws-cdk/toolkit-lib/lib/api/drift/drift-formatter.ts:
```diff
@@ -0,0 +1,283 @@
+  /**
+   * Format the stack drift detection results
+   */
+  public formatStackDrift(): DriftResult {
```
Please keep internal types and public types separate

Commit:
Several reverts

Leo10Gama committed:
Merge branch 'main' into drift

Commit:
Merge branch 'drift' of https://github.com/Leo10Gama/aws-cdk-cli into drift

Commit:
Move drift integ tests to folder

Commit:
Fix import issues with moving tests

mrgrain commented on packages/@aws-cdk/toolkit-lib/lib/toolkit/toolkit.ts:
```diff
@@ -375,6 +378,85 @@ export class Toolkit extends CloudAssemblySourceBuilder {
+
+      const driftOutput = formatter.formatStackDrift();
+
+      if (driftOutput.formattedDrift.stackHeader) {
+        await ioHelper.notify(IO.CDK_TOOLKIT_I4500.msg(driftOutput.formattedDrift.stackHeader, driftOutput));
+        output.push(driftOutput.formattedDrift.stackHeader);
+      }
+      if (driftOutput.formattedDrift.unchanged) {
+        await ioHelper.notify(IO.CDK_TOOLKIT_I4501.msg(driftOutput.formattedDrift.unchanged, driftOutput));
+        output.push(driftOutput.formattedDrift.unchanged);
+      }
+      if (driftOutput.formattedDrift.unchecked) {
+        await ioHelper.notify(IO.CDK_TOOLKIT_I4502.msg(driftOutput.formattedDrift.unchecked, driftOutput));
+        output.push(driftOutput.formattedDrift.unchecked);
+      }
+      if (driftOutput.formattedDrift.modified) {
+        await ioHelper.notify(IO.CDK_TOOLKIT_I4503.msg(driftOutput.formattedDrift.modified, driftOutput));
+        output.push(driftOutput.formattedDrift.modified);
+      }
+      if (driftOutput.formattedDrift.deleted) {
+        await ioHelper.notify(IO.CDK_TOOLKIT_I4504.msg(driftOutput.formattedDrift.deleted, driftOutput));
+        output.push(driftOutput.formattedDrift.deleted);
+      }
+      if (driftOutput.formattedDrift.finalResult) {
+        await ioHelper.notify(IO.CDK_TOOLKIT_I4500.msg(driftOutput.formattedDrift.finalResult, driftOutput));
+        output.push(driftOutput.formattedDrift.finalResult);
+      }
+      drifts += driftOutput.numResourcesWithDrift === -1 ? 0 : driftOutput.numResourcesWithDrift || 0;
+      uncheckedResources += driftOutput.numResourcesUnchecked === -1 ? 0 : driftOutput.numResourcesUnchecked || 0;
+      output.push(driftOutput.formattedDrift);
+    }
```
We don't need to send the same payload multiple times and also it's the same stuff as in the message already.

Header and footer (finalResult?) can just be default messages without code and payload, they are not super interesting. Let's just turn them all into default messages.

Leo10Gama replied:
When you say default messages, are you referring to making `CDK_TOOLKIT_I4500` take no payload? The default messages seem to have been removed from `IO`

Leo10Gama replied:
Nevermind, discovered it as `ioHelper.defaults.<loggingLevel>`

Commit:
Refactor

Commit:
Cleanup messges

Commit:
Fixnaming converntions

Commit:
Revise tests to use the TestIoHelper

Commit:
I think that fixed the licensing??

Commit:
Revert some package schenanigans

Commit:
Licensing fixed this time I think

Leo10Gama commented on packages/@aws-cdk/integ-runner/THIRD_PARTY_LICENSES:
Honestly not sure how this came about 😅 was a result of the `yarn build` script, and I'm not sure how to get it fixed

rix0rrr committed:
Merge remote-tracking branch 'origin/main' into pr/Leo10Gama/442

Commit:
Fix IoHelper import

Commit:
Fix license

Leo10Gama committed:
Merge branch 'main' into drift

mrgrain commented on packages/@aws-cdk/toolkit-lib/lib/toolkit/toolkit.ts:
```diff
@@ -375,6 +378,79 @@ export class Toolkit extends CloudAssemblySourceBuilder {
+        output.push(driftOutput.sections?.stackHeader);
+      }
+      if (driftOutput.sections?.unchanged) {
+        await ioHelper.notify(IO.CDK_TOOLKIT_I4500.msg(driftOutput.sections?.unchanged));
```
Either default message or a payload

mrgrain commented on packages/@aws-cdk/toolkit-lib/lib/api/drift/drift-formatter.ts:
```diff
@@ -0,0 +1,305 @@
+      output.push(finalResult);
+      return {
+        numResourcesWithDrift: 0,
+        numResourcesUnchecked: this.allStackResources ? this.allStackResources.size - this.driftResults.StackResourceDrifts.length : -1,
```
```suggestion
        numResourcesUnchecked: this.allStackResources ? this.allStackResources.size - this.driftResults.StackResourceDrifts.length : undefined,
```

mrgrain commented on packages/@aws-cdk/toolkit-lib/lib/toolkit/toolkit.ts:
```diff
@@ -375,6 +378,79 @@ export class Toolkit extends CloudAssemblySourceBuilder {
+      );
+
+      // Get all stack resources if we want to show all
+      const allStackResources = new Map<string, string>();
+      Object.keys(stack.template.Resources || {}).forEach(id => {
+        const resource = stack.template.Resources[id];
+        allStackResources.set(id, resource.Type);
+      });
```
This should be movable to inside the DriftFormatter

mrgrain commented on packages/@aws-cdk/toolkit-lib/lib/actions/drift/index.ts:
```diff
@@ -0,0 +1,62 @@
+   */
+  readonly numResourcesUnchecked?: number;
+
+  /**
+   * Formatted sections of the drift results
+   */
+  readonly sections?: FormattedDrift;
+
+  /**
+   * Complete formatted drift
+   */
+  readonly formattedDrift: string;
```
```suggestion
  /**
   * Complete formatted drift
   */
  readonly formattedDrift: FormattedDrift;
```

mrgrain commented on packages/@aws-cdk/toolkit-lib/lib/actions/drift/index.ts:
```diff
@@ -0,0 +1,62 @@
+  readonly stacks: StackSelector;
+}
+
+export interface FormattedDrift {
+  /**
+   * Resources that have not changed
+   */
+  readonly unchanged?: string;
+
+  /**
+   * Resources that were not checked for drift
+   */
+  readonly unchecked?: string;
+
+  /**
+   * Resources with drift
+   */
+  readonly modified?: string;
+
+  /**
+   * Resources that have been deleted (drift)
+   */
+  readonly deleted?: string;
+
+  /**
+   * The header, containing the stack name
+   */
+  readonly stackHeader?: string;
+
+  /**
+   * The final results (summary) of the drift results
+   */
+  readonly finalResult?: string;
+}
```
```suggestion
export interface FormattedDrift {
  /**
   * Resources that have not changed
   */
  readonly unchanged?: string;

  /**
   * Resources that were not checked for drift
   */
  readonly unchecked?: string;

  /**
   * Resources with drift
   */
  readonly modified?: string;

  /**
   * Resources that have been deleted (drift)
   */
  readonly deleted?: string;
}
```

mrgrain commented on packages/@aws-cdk/toolkit-lib/lib/toolkit/toolkit.ts:
```diff
@@ -375,6 +378,79 @@ export class Toolkit extends CloudAssemblySourceBuilder {
+    return {
+      numResourcesWithDrift: drifts,
+      numResourcesUnchecked: uncheckedResources,
+      formattedDrift: output.join('\n\n'),
```
```suggestion
      formattedDrift: {
        unchanged: driftOutput. sections?.unchanged,
        unchecked: driftOutput. sections?.unchecked,
        modified: driftOutput. sections?.modified,
        deleted: driftOutput. sections?.deleted,
      },
```

mrgrain commented on packages/@aws-cdk/toolkit-lib/lib/toolkit/toolkit.ts:
```diff
@@ -375,6 +378,79 @@ export class Toolkit extends CloudAssemblySourceBuilder {
+        await ioHelper.defaults.result(driftOutput.sections?.finalResult);
+        output.push(driftOutput.sections?.finalResult);
+      }
+      drifts += driftOutput.numResourcesWithDrift === -1 ? 0 : driftOutput.numResourcesWithDrift || 0;
+      uncheckedResources += driftOutput.numResourcesUnchecked === -1 ? 0 : driftOutput.numResourcesUnchecked || 0;
```
not -1

mrgrain commented on packages/@aws-cdk/toolkit-lib/lib/api/drift/drift.ts:
```diff
@@ -0,0 +1,84 @@
+      await ioHelper.defaults.trace('Waiting for drift detection to complete...');
+      checkIn = Date.now() + timeBetweenOutputs;
+    }
+    await new Promise(resolve => setTimeout(resolve, 500));
```
```suggestion
    const timeBetweenApiCall = 1000;
    await new Promise(resolve => setTimeout(resolve, timeBetweenApiCall));
```

Commit:
Implement requested changes

Leo10Gama committed:
Merge branch 'main' into drift

Commit:
Fix testing errors

Commit:
Minor change

mrgrain committed:
Merge branch 'main' into drift

mrgrain committed:
fixes

mrgrain committed:
test fixes

mrgrain committed:
fixup integ test
